### PR TITLE
Unwrap basic token parsers

### PR DESCRIPTION
**Pull request by Kha**
@leodemoura I successfully applied a workaround for the dreaded "changing syntax already being used in syntax quotations" issue: temporarily marking the changes as being active only inside syntax quotations. Recall that the fundamental issue is that when we change syntax, the resulting binary will still have the old syntax structure embedded from syntax quotations. So it would have to be able to handle (in the elaborator etc.) both this old syntax and the new syntax produced by its parser. We previously discussed solving this issue by parsing syntax quotations using the parsers of the current stage via the interpreter. But a much simpler solution is to activate the new syntax inside syntax quotations one stage earlier than outside, so that from the point of view of the elaborator etc., they change at the same time.

In other words:
* stage N (embedded: old syntax, produces: old syntax, elaborates: old syntax):
  * add new syntax guarded by `checkInsideQuot`
* stage N+1 (embedded: old syntax, produces: new syntax inside quotations, elaborates: old syntax):
  * adjust syntax consumers to new syntax
  * remove syntax guards and old syntax
* stage N+2 (embedded: new syntax, produces: new syntax, elaborates: new syntax):
  * done!

Note that this does not handle correctly nested quotations, where we should first activate the syntax for the innermost quotations, then the surrounding one, then... But I doubt this will be an issue in practice.

### CODE REVIEW

File: src/Lean/Parser.lean [hunk 1]
```diff
@@ -12,12 +12,21 @@ import Lean.Parser.Module
...
+def mkAntiquot.parenthesizer (name : String) (kind : Option SyntaxNodeKind) (anonymous := true) : PrettyPrinter.Parenthesizer :=
+Parser.mkAntiquot.parenthesizer name kind anonymous
 
-end Parser
+-- The parenthesizer auto-generated these instances correctly, but tagged them with the wrong kind, since the actual kind
+-- (e.g. `ident`) is not equal to the parser name `Lean.Parser.Term.ident`.
+@[builtinParenthesizer ident] def ident.parenthesizer := Parser.Term.ident.parenthesizer
+@[builtinParenthesizer numLit] def numLit.parenthesizer := Parser.Term.num.parenthesizer
+@[builtinParenthesizer charLit] def charLit.parenthesizer := Parser.Term.char.parenthesizer
+@[builtinParenthesizer strLit] def strLit.parenthesizer := Parser.Term.str.parenthesizer
```
Comment (Kha) [inline src/Lean/Parser.lean:28]: Note that here we basically say "we want all identifiers pretty printed like identifier *terms* are", which I suppose is fine in practice. Analogous formatter code will be added here once I refactor it to be precompiled.

Comment (Kha) [reply src/Lean/Parser.lean:28]: Users who want to have identifiers etc. behave differently in their own syntax category can always add a wrapper node specifically for that category.

File: src/Init/LeanInit.lean [hunk 2]
```diff
@@ -223,12 +223,12 @@ def isOfKind : Syntax → SyntaxNodeKind → Bool
...
 match stx with
 | Syntax.node _ args => args.get! i
-| _                  => arbitrary _
+| _                  => panic! "Syntax.getArg: not a node"
 
 def getArgs (stx : Syntax) : Array Syntax :=
 match stx with
 | Syntax.node _ args => args
-| _                  => #[]
+| _                  => panic! "Syntax.getArgs: not a node"
```
Comment (Kha) [inline src/Init/LeanInit.lean:231]: Do you agree with this change? It definitely makes debugging much easier.

Comment (Kha) [reply src/Init/LeanInit.lean:231]: The names could be changed to reflect the possibility of panics, thought this was already true for `getArg`.

Comment (leodemoura) [reply src/Init/LeanInit.lean:231]: These changes are problematic. Users may create buggy macros that produce invalid `Syntax` objects for the builtin elaboration functions. We use `getArg` and `getArgs` all over the place in the elaborator.

**Comment by leodemoura** [comment]
Merging as-is.
